crates/fuel-core/src/query/balance/asset_query.rs

```diff
@@ -196,7 +213,14 @@ impl<'a> AssetQuery<'a> {
 
     /// Returns the iterator over all valid(spendable, allowed by `exclude`) coins of the `owner`
     /// for the `asset_id`.
-    pub fn coins(&self) -> impl Iterator<Item = StorageResult<CoinType>> + '_ {
+    pub fn coins(self) -> impl Stream<Item = StorageResult<CoinType>> + 'a {
         self.query.coins()
     }
 }
+
+fn has_asset(assets: &Option<HashSet<&AssetId>>, asset_id: &AssetId) -> bool {
```

Comment:
hmm, maybe the signature of this function should be -
```suggestion
fn has_asset(assets: &HashSet<&AssetId>, asset_id: &AssetId) -> bool {
```

and leave it to the caller to decide whether to call or not based on `assets` having a value inside the `Option`.

Reply:
https://github.com/FuelLabs/fuel-core/pull/2341/commits/c355cd1569c785379018912aa3ab1a1661d7ee8b